prow/plugins/approve/approve.go

```diff
@@ -730,3 +663,241 @@ func getLast(cs []*comment) *comment {
+		sa.approversHandler.AddApprover(
+			c.Author,
+			c.HTMLURL,
+			false,
+		)
+	}
+	if reviewActsAsApprove && c.ReviewState == github.ReviewStateChangesRequested {
+		sa.approversHandler.RemoveApprover(c.Author)
+	}
+
+	for _, match := range commandRegex.FindAllStringSubmatch(c.Body, -1) {
+		name := strings.ToUpper(match[1])
+		if name == removeApproveCommand {
+			sa.approversHandler.RemoveApprover(c.Author)
+			continue
+		}
+		if name != approveCommand && name != lgtmCommand {
+			continue
+		}
+		args := strings.ToLower(strings.TrimSpace(match[2]))
+		if strings.Contains(args, cancelArgument) {
+			sa.approversHandler.RemoveApprover(c.Author)
+			continue
+		}
+
+		if c.Author == author {
+			sa.approversHandler.AddAuthorSelfApprover(
+				c.Author,
+				c.HTMLURL,
+				args == noIssueArgument,
+			)
+		}
+
+		if name == approveCommand {
+			sa.approversHandler.AddApprover(
+				c.Author,
+				c.HTMLURL,
+				args == noIssueArgument,
+			)
+		} else {
+			sa.approversHandler.AddLGTMer(
+				c.Author,
+				c.HTMLURL,
+				args == noIssueArgument,
+			)
+		}
+
+	}
+}
+
+func (sa *simpleApproveHelper) isApproved() bool {
+	return sa.approversHandler.IsApproved()
+}
+
+func (sa *simpleApproveHelper) updateNoficiation(linkURL *url.URL, commandHelpLink, prProcessLink, org, repo, branch string, latestNotification *comment) *string {
+	message := approvers.GetMessage(sa.approversHandler, linkURL, commandHelpLink, prProcessLink, org, repo, branch)
+	if message == nil || (latestNotification != nil && strings.Contains(latestNotification.Body, *message)) {
+		return nil
+	}
+	return message
+}
+
+var _ approveHelper = &granularApproveHelper{}
+
+type granularApproveHelper struct {
+	approversHandler approvers2.Approvers
+}
+
+func (ga *granularApproveHelper) processSelfApproval(approve bool, author string, url string) {
+	if approve {
+		ga.approversHandler.AddAuthorSelfApprover(author, url, "")
+	} else {
+		ga.approversHandler.AddAssignees(author)
+	}
+}
+
+func (ga *granularApproveHelper) addAssignees(logins ...string) {
+	ga.approversHandler.AddAssignees(logins...)
+}
+
+func (ga *granularApproveHelper) setAssociatedIssue(issue int) {
+	ga.approversHandler.AssociatedIssue = issue
+}
+
+func (ga *granularApproveHelper) setRequiredIssue(required bool) {
+	ga.approversHandler.RequireIssue = required
+}
+
+func (ga *granularApproveHelper) setManuallyApproved(f func() bool) {
+	ga.approversHandler.ManuallyApproved = f
+}
+
+func (ga *granularApproveHelper) processComment(c *comment, author string, reviewActsAsApprove bool) {
+	if c.Author == "" {
+		return
+	}
+
+	if reviewActsAsApprove && c.ReviewState == github.ReviewStateApproved {
+		ga.approversHandler.AddApprover(
+			c.Author,
+			c.HTMLURL,
+			"",
+		)
+	}
+	if reviewActsAsApprove && c.ReviewState == github.ReviewStateChangesRequested {
+		ga.approversHandler.RemoveApprover(c.Author)
+	}
+
+	for _, match := range commandRegex.FindAllStringSubmatch(c.Body, -1) {
```

Comment:
Since the approvals in the new approve logic are appended (a user can give multiple approvals approving different sections of the PR) the order or processing comments and github reviews does not matter.

If a user submit a github review with "approve" state it just acts as a blanket approval. We should document this behavior in the plugin docs. (Here: https://github.com/kubernetes-sigs/prow/pull/16/files ?)

Reply:
>If a user submit a github review with "approve" state it just acts as a blanket approval. We should document this behavior in the plugin docs.

No.  The approve plugin doesn't work this way today.  Note "approval commands override GitHub review state" https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approve.go#L243-L249

As Ben said, it would be unintuitive if turning on `granular_approval` changed this behavior.

If I submit an "approved" review with a comment of `/approve files foo`, it should only approve `foo`, and not be treated as a blanket approval

Reply:
@spiffxp -- working on the remaining comments, but wanted some clarification on this one.

The logic here is the same as the current approve plugin. Instead of processing the logic in `addApprovers` as here:
https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approve.go#L586

the logic is refactored into a new function called `processComment`. If at all an approver gives an "approve review" with a `/approve files foo` comment, the `handleReviewEvent` event handler should let the `genericCommentEventHandler` handle it like you linked above. 

I can add a test for this case in the follow-up commits too.
Please lmk if I'm misunderstanding something here.